cpp/include/Ice/ClientAuthenticationOptions.h

```diff
@@ -0,0 +1,170 @@
+#    include <sspi.h>
+#    undef SECURITY_WIN32
+#elif defined(__APPLE__)
+#    include <Security/SecureTransport.h>
+#    include <Security/Security.h>
+#else
+#    include <openssl/ssl.h>
+#endif
+
+namespace Ice::SSL
+{
+    /**
+     * The SSL configuration properties for client connections.
+     */
+    struct ClientAuthenticationOptions
+    {
+#if defined(_WIN32)
+        /**
+         * A callback that allows selecting the client's credentials based on the target server host name.
+         *
+         * @param host The target server host name.
+         * @return The client credentials. The credentials must remain valid for the duration of the connection.
+         *
+         * [See Detailed Schannel documentation on Schannel credentials](
+         * https://learn.microsoft.com/en-us/windows/win32/secauthn/acquirecredentialshandle--schannel)
+         */
+        std::function<CredHandle(const std::string& host)> clientCredentialsSelectionCallback;
+
+        /**
+         * A callback that allows manually validating the server certificate chain. When the verification callback
+         * returns false, the connection will be aborted with an Ice::SecurityException.
+         *
+         * @param context A CtxtHandle representing the security context associated with the current connection. This
+         * context contains security data relevant for validation, such as the server's certificate chain and cipher
+         * suite.
+         * @param info The IceSSL::ConnectionInfoPtr object that provides additional connection-related data
+         * which might be relevant for contextual certificate validation.
+         * @return true if the certificate chain is valid and the connection should proceed; false if the certificate
+         * chain is invalid and the connection should be aborted. An exception may be thrown to indicate a custom
+         * error during the validation process.
+         *
+         * [See Manually Validating Schannel Credentials for more
+         * details](https://learn.microsoft.com/en-us/windows/win32/secauthn/manually-validating-schannel-credentials).
+         */
+        std::function<bool(CtxtHandle context, const IceSSL::ConnectionInfoPtr& info)>
+            serverCertificateValidationCallback;
+#elif defined(__APPLE__)
+        /**
+         * A callback that allows selecting the client's certificate chain based on the target server host name.
+         *
+         * @param host The target server host name.
+         * @return The client's certificate chain. The certificate chain must remain valid for the duration of the
+         * connection.
+         *
+         * The requirements for the Secure Transport certificate chain are documented in
+         * https://developer.apple.com/documentation/security/1392400-sslsetcertificate?changes=_3&language=objc
+         */
+        std::function<CFArrayRef(const std::string& host)> clientCertificateSelectionCallback;
+
+        /**
```

Comment:
It's not clear how trustedRootCertificates interacts with serverCertificateValidationCallback.

trustedRootCertificates is used only when serverCertificateValidationCallback is null?

Reply:
The trusted certificates are only used when the validation callback is null.

Reply:
Reworked this and updated the docs.